Fix high CPU usage for long SVG text workloads

nullhook opened the pull request:
#### 2bd61e2f1c05935f2600b9c0edac21f60783ce39
<pre>
Fix high CPU usage for long SVG text workloads
<a href="https://bugs.webkit.org/show_bug.cgi?id=291574">https://bugs.webkit.org/show_bug.cgi?id=291574</a>
<a href="https://rdar.apple.com/147359331">rdar://147359331</a>

Reviewed by Vitor Roriz.

SVGTextMetricsBuilder advances character by character,
instantiating SVGTextMetrics and calling FontCascade::width, which
creates a ComplexController. This occurs in a busy loop, creating and
discarding many ComplexControllers, which is expensive.

This optimization reuses a single ComplexController within the busy loop
to handle all complex text advances.

* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
(WebCore::SVGTextMetricsBuilder::advanceComplexText):
(WebCore::SVGTextMetricsBuilder::measureTextRenderer):
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.h:

Canonical link: <a href="https://commits.webkit.org/293774@main">https://commits.webkit.org/293774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c031521c9b4039d7cfc9589f2b410485f8f4e1c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99897 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19545 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9829 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105024 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50478 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101938 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19850 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27980 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76046 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33138 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102904 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15140 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90210 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56404 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14944 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8195 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49847 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84870 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8284 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107385 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27010 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19735 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84998 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27372 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86410 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84521 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21465 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29198 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6921 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20829 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26947 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32163 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26758 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30074 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28317 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
```diff
@@ -76,8 +76,19 @@ void SVGTextMetricsBuilder::advanceSimpleText()
 void SVGTextMetricsBuilder::advanceComplexText()
 {
     unsigned metricsLength = currentCharacterStartsSurrogatePair() ? 2 : 1;
```
Do we need this? I think ComplexTextController should know how to handle surrogated pairs. Shouldn't we always pass 1 here?

For example, `adjustSelectionRectForComplexText` uses a `ComplexTextController` to advance to advance in regarding the indexes on a text run without having to handle surrogated pairs by itself.

nullhook replied:
if surrogate pairs are handled internally by ComplexTextController, how do we determine the logical length it has advanced? the value (either 1 or 2) is utilized when storing in SVGTextMetrics

vitorroriz commented on Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
```diff
@@ -76,8 +76,19 @@ void SVGTextMetricsBuilder::advanceSimpleText()
+    offsetEndRange = m_complexTextController->runWidthSoFar();
+
+    m_currentMetrics = SVGTextMetrics(*m_text, metricsLength, offsetEndRange - offsetStartRange);
+    m_complexStartToCurrentMetrics = SVGTextMetrics(*m_text, m_textPosition + metricsLength, offsetEndRange);
```
Your patch changes which `SVGTextMetrics` constructor we use. A weird thing (not introduced by this patch) is that the former constructor (invoked by `measureCharacterRange`) initializes `m_glyph`, but the constructor used now doesn't. For example, it assigns a `unicodeString` to `m_glyph` which comes from the TextRun that was build using `SVGTextMetrics::constructTextRun`. 

The constructor we are using right now does not initialize `m_glyph`. Is this glyph data used somewhere? I see this comment from @nikolaszimmermann :
```
    // Only useful when measuring individual characters, to lookup ligatures.
    const Glyph& glyph() const { return m_glyph; }
 ```
 
 However, I think I don't see any references to `m_glyph` or `glyph()`, apart from `isEmpty` which does not matter here, since we are still setting `width`.
 
 Maybe @nikolaszimmermann would know something about it.

nikolaszimmermann replied:
The new constructor was added 1 yr ago, in https://commits.webkit.org/277214@main, not caring about m_glyph -- which would be a bug in the first place, if it wouldn't be unused nowadays.

The whole logic was present to support SVG kerning information (`<hkern>` / `<vkern>`) - a relict from the past. The last user of `glyph()` got removed in 2016:

```
commit 4738cfa45c28bf60411bd29119e154f4bae7d963
Author: Myles C. Maxfield <mmaxfield@apple.com>
Date:   Sun Mar 13 00:36:59 2016 +0000

    Delete dead SVG Font code
    https://bugs.webkit.org/show_bug.cgi?id=154718
    
    Reviewed by Antti Koivisto.
```.

So you can go ahead and nuke m_glyph & co.

nikolaszimmermann reviewed (commented):
Nice approach -- what tests are you using to evaluate performance?

litherum commented on Source/WebCore/rendering/svg/SVGTextMetricsBuilder.h:
```diff
@@ -57,6 +58,7 @@ class SVGTextMetricsBuilder {
 
     // Simple text only.
     std::unique_ptr<WidthIterator> m_simpleWidthIterator;
+    std::unique_ptr<ComplexTextController> m_complexTextController;
```
seems like this should not be in the `// Simple text only.` section but should instead be in the `// Complex text only.` section below

litherum commented on Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
```diff
@@ -76,8 +76,18 @@ void SVGTextMetricsBuilder::advanceSimpleText()
+    float beforeWidth = 0;
+    float afterWidth = 0;
+
+    m_complexTextController->advance(m_textPosition, nullptr);
```
Are you sure you don't need to check if `m_textPosition` is behind the current position of the complex text controller? The previous code was stateless, and you're introducing state into it now.

nullhook replied:
so, the controller resets its internal states if the passed offset is behind the current position. i'm assuming we always want the logic here to be monotonic? i'm curious when `m_textPosition` would be behind. would it be an unintended regression?

btw, this controller doesn't seem to support an API to read its current position. are users of this controller expected to maintain current positions themselves?

vitorroriz reviewed (changes requested):
It seems we are breaking related SVG tests for WPE/GTK, we probably want to understand why that's the case.

vitorroriz commented on LayoutTests/svg/text/crash-long-text-content.html:
Is this really a "crash test"?

vitorroriz replied:
Is this the minimum reduced content for reproducing the hanging behavior?

vitorroriz replied:
I guess you will need to skip these tests for GTK/WPE platforms (on TestExpectation) as we are not implementing the fix for them.

nullhook replied:
yes, it's minimum content to repro timeout. added tests to skip on gtk/wpe.

vitorroriz commented on Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
```diff
@@ -76,9 +76,26 @@ void SVGTextMetricsBuilder::advanceSimpleText()
 {
     unsigned metricsLength = currentCharacterStartsSurrogatePair() ? 2 : 1;
+
+    // FIXME: NON-COCOA platforms shifts some pixels for SVGText when x,y, and rotate are used.
```
You can reference the bugzilla you have created for GTK/WPE in this FIXME. There is an useful short url format: `webkit.org/b/<bugzilla-id>`

vitorroriz commented on Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
```diff
@@ -76,9 +76,27 @@ void SVGTextMetricsBuilder::advanceSimpleText()
+#else
     m_currentMetrics = SVGTextMetrics::measureCharacterRange(*m_text, m_textPosition, metricsLength);
     m_complexStartToCurrentMetrics = SVGTextMetrics::measureCharacterRange(*m_text, 0, m_textPosition + metricsLength);
-    ASSERT(m_currentMetrics.length() == metricsLength);
```
I guess we have no reason for deleting this ASSERT for the other ports

nullhook replied:
added back

nullhook committed:
Fix high CPU usage for long SVG text workloads
https://bugs.webkit.org/show_bug.cgi?id=291574
rdar://147359331

Reviewed by Vitor Roriz.

SVGTextMetricsBuilder advances character by character,
instantiating SVGTextMetrics and calling FontCascade::width, which
creates a ComplexController. This occurs in a busy loop, creating and
discarding many ComplexControllers, which is expensive.

This optimization reuses a single ComplexController within the busy loop
to handle all complex text advances.

* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
(WebCore::SVGTextMetricsBuilder::advanceComplexText):
(WebCore::SVGTextMetricsBuilder::measureTextRenderer):
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.h:

Canonical link: https://commits.webkit.org/293774@main
